File: sim/btd/upgrades.py
"""Tower upgrade definitions. Ports BloonsTD.GetUpgrade() from BloonsTD.as.

Each tower has up to 4 upgrades. Two paths of two upgrades each:
  path 1 = upgrade1 (level 1) then upgrade3 (level 2)
  path 2 = upgrade2 (level 1) then upgrade4 (level 2)

Path-locking (from AS clickUpgradeBtn):
  - The upgrade1 button is blocked once upgrade2 is bought, which means you
    can never buy upgrade1 or upgrade3 after committing to path 2. Path 2 has
    no equivalent block in the other direction, so buying path 1 first and
    then progressing to path 2 is allowed.

Out of scope (omitted from the table; depend on monkey storm consumable which
the first RL iteration excludes):
  - beacon3 (storm unlock)
  - beacon4 (activate monkey storm)

Some upgrades have behavioural side-effects beyond stat changes; those are
documented on the spec and implemented inline in game.py:
  - bomb upgrade2: bombs spawn N frag bullets on detonation
  - ice  upgrade2: permafrost — frozen bloons emerge at half speed
  - ice  upgrade4: snap freeze — 39% chance to pop on freeze
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class UpgradeSpec:
    cost: int
    # Stat deltas applied with +=.
    additive: dict[str, float] = field(default_factory=dict)
    # Stat values applied with =.
    absolute: dict[str, Any] = field(default_factory=dict)
    # Boolean flags flipped to True.
    flags: dict[str, bool] = field(default_factory=dict)
    # Transform upgrades (blade, glaive, missile, multishot) reset the firing
    # cooldown so the tower fires immediately after the upgrade.
    reset_tsls: bool = False


UPGRADES: dict[str, UpgradeSpec] = {
    # Dart Monkey — symmetric range / pierce on both paths.
    "dart1": UpgradeSpec(cost=90, additive={"attack_radius": 25}),
    "dart2": UpgradeSpec(cost=90, additive={"attack_radius": 25}),
    "dart3": UpgradeSpec(cost=140, additive={"pierce_max": 1}),
    "dart4": UpgradeSpec(cost=120, additive={"pierce_max": 1}),

    # Tack Shooter.
    "tack1": UpgradeSpec(cost=200, additive={"attack_rate": -15}),
    "tack2": UpgradeSpec(cost=180, additive={"attack_rate": -5},
                         flags={"transformed": True}, reset_tsls=True),
    "tack3": UpgradeSpec(cost=100, additive={"attack_radius": 10},
                         absolute={"bullet_scale": 1.3}),
    "tack4": UpgradeSpec(cost=100, additive={"attack_radius": 10},
                         absolute={"bullet_scale": 1.5}),

    # Boomerang.
    "boomerang1": UpgradeSpec(cost=270, additive={"pierce_max": 3}),
    "boomerang2": UpgradeSpec(cost=280, additive={"pierce_max": 3},
                              flags={"transformed": True}, reset_tsls=True),
    "boomerang3": UpgradeSpec(cost=150, flags={"icebreak": True}),
    "boomerang4": UpgradeSpec(cost=120, flags={"leadbreak": True}),

    # Bomb (Cannon). upgrade2 spawns frag bullets on detonation (handled in
    # game._on_hit).
    "bomb1": UpgradeSpec(cost=430, absolute={"bullet_scale": 1.5}),
    "bomb2": UpgradeSpec(cost=220),
    "bomb3": UpgradeSpec(cost=200, additive={"attack_radius": 20}),
    "bomb4": UpgradeSpec(cost=210, additive={"attack_rate": -8},
                         absolute={"shoot_power": 25.0},
                         flags={"transformed": True}, reset_tsls=True),

    # Ice Ball. upgrade2 permafrost / upgrade4 snap freeze are flag-only;
    # effects live in game._try_freeze and game._tick_bloons.
    "ice1": UpgradeSpec(cost=250, additive={"freeze_len": 20}),
    "ice2": UpgradeSpec(cost=250),
    "ice3": UpgradeSpec(cost=200, additive={"attack_radius": 15},
                        absolute={"bullet_scale": 1.0}),
    "ice4": UpgradeSpec(cost=290),

    # Super Monkey. Laser/plasma flip icebreak/leadbreak via the flags;
    # `laser=True` is kept as a marker (visualisation hook for future).
    "super1": UpgradeSpec(cost=1000, additive={"attack_radius": 50}),
    "super2": UpgradeSpec(cost=1400, additive={"attack_radius": 50}),
    "super3": UpgradeSpec(cost=3500, additive={"pierce_max": 1},
                          flags={"icebreak": True, "laser": True}),
    "super4": UpgradeSpec(cost=4000, additive={"pierce_max": 1},
                          absolute={"attack_rate": 1},
                          flags={"icebreak": True, "leadbreak": True, "laser": True}),

    # Spike-o-pult.
    "spikeopult1": UpgradeSpec(cost=250, additive={"attack_radius": 20}),
    "spikeopult2": UpgradeSpec(cost=825, absolute={"pierce_max": 20}),
    "spikeopult3": UpgradeSpec(cost=250, additive={"attack_rate": -8}),
    "spikeopult4": UpgradeSpec(cost=575, flags={"transformed": True, "is_spread": True},
                               reset_tsls=True),

    # Beacon. Drums (upgrade2) toggles the rate buff via beacon-refresh path.
    "beacon1": UpgradeSpec(cost=500, additive={"attack_radius": 30}),
    "beacon2": UpgradeSpec(cost=1500),
    # beacon3, beacon4 omitted (monkey storm — deferred).
}


def next_path_upgrade(tower_type: str, path: int,
                      u1: bool, u2: bool, u3: bool, u4: bool) -> str | None:
    """Return the upgrade name the given path button would buy next, or None
    if the path is exhausted or locked. Path 1 = upgrade1 then upgrade3,
    Path 2 = upgrade2 then upgrade4. Path 1 is locked once upgrade2 is bought."""
    if path == 1:
        if u2:
            return None
        if not u1:
            return f"{tower_type}1"
        if not u3:
            return f"{tower_type}3"
        return None
    if path == 2:
        if u4:
            return None
        if not u2:
            return f"{tower_type}2"
        name = f"{tower_type}4"
        # beacon4 is not a real upgrade (monkey-storm action); refuse.
        return name if name in UPGRADES else None
    raise ValueError(f"path must be 1 or 2, got {path}")

File: sim-rs/src/upgrades.rs
/// Tower upgrade definitions. Ports BloonsTD.GetUpgrade() and sim/btd/upgrades.py.
///
/// Two paths of two upgrades each:
///   Path 1: upgrade1 -> upgrade2  (flags upgrade1, upgrade2)
///   Path 2: upgrade3 -> upgrade4  (flags upgrade3, upgrade4)
/// Paths are fully independent — both can be maxed in any interleaved order.

use crate::constants::TowerType;

#[derive(Clone, Debug)]
pub(crate) struct UpgradeSpec {
    pub cost: u32,
    /// Stat deltas applied with +=.
    pub additive: &'static [(&'static str, f64)],
    /// Stat values applied with =.
    pub absolute: &'static [(&'static str, f64)],
    /// Boolean flags flipped to True.
    pub flags: &'static [(&'static str, bool)],
    /// Transform upgrades reset the firing cooldown.
    pub reset_tsls: bool,
}

// All upgrades keyed by name. Same structure as Python's UPGRADES dict.
pub(crate) const UPGRADES: &[(&str, UpgradeSpec)] = &[
    // Dart Monkey
    ("dart1", UpgradeSpec { cost: 90, additive: &[("attack_radius", 25.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("dart2", UpgradeSpec { cost: 90, additive: &[("attack_radius", 25.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("dart3", UpgradeSpec { cost: 140, additive: &[("pierce_max", 1.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("dart4", UpgradeSpec { cost: 120, additive: &[("pierce_max", 1.0)], absolute: &[], flags: &[], reset_tsls: false }),

    // Tack Shooter
    ("tack1", UpgradeSpec { cost: 200, additive: &[("attack_rate", -15.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("tack2", UpgradeSpec { cost: 180, additive: &[("attack_rate", -5.0)], absolute: &[], flags: &[("transformed", true)], reset_tsls: true }),
    ("tack3", UpgradeSpec { cost: 100, additive: &[("attack_radius", 10.0)], absolute: &[("bullet_scale", 1.3)], flags: &[], reset_tsls: false }),
    ("tack4", UpgradeSpec { cost: 100, additive: &[("attack_radius", 10.0)], absolute: &[("bullet_scale", 1.5)], flags: &[], reset_tsls: false }),

    // Boomerang
    ("boomerang1", UpgradeSpec { cost: 270, additive: &[("pierce_max", 3.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("boomerang2", UpgradeSpec { cost: 280, additive: &[("pierce_max", 3.0)], absolute: &[], flags: &[("transformed", true)], reset_tsls: true }),
    ("boomerang3", UpgradeSpec { cost: 150, additive: &[], absolute: &[], flags: &[("icebreak", true)], reset_tsls: false }),
    ("boomerang4", UpgradeSpec { cost: 120, additive: &[], absolute: &[], flags: &[("leadbreak", true)], reset_tsls: false }),

    // Bomb (Cannon)
    ("bomb1", UpgradeSpec { cost: 430, additive: &[], absolute: &[("bullet_scale", 1.5)], flags: &[], reset_tsls: false }),
    ("bomb2", UpgradeSpec { cost: 220, additive: &[], absolute: &[], flags: &[], reset_tsls: false }),
    ("bomb3", UpgradeSpec { cost: 200, additive: &[("attack_radius", 20.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("bomb4", UpgradeSpec { cost: 210, additive: &[("attack_rate", -8.0)], absolute: &[("shoot_power", 25.0)], flags: &[("transformed", true)], reset_tsls: true }),

    // Ice Ball
    ("ice1", UpgradeSpec { cost: 250, additive: &[("freeze_len", 20.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("ice2", UpgradeSpec { cost: 250, additive: &[], absolute: &[], flags: &[], reset_tsls: false }),
    ("ice3", UpgradeSpec { cost: 200, additive: &[("attack_radius", 15.0)], absolute: &[("bullet_scale", 1.0)], flags: &[], reset_tsls: false }),
    ("ice4", UpgradeSpec { cost: 290, additive: &[], absolute: &[], flags: &[], reset_tsls: false }),

    // Super Monkey
    ("super1", UpgradeSpec { cost: 1000, additive: &[("attack_radius", 50.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("super2", UpgradeSpec { cost: 1400, additive: &[("attack_radius", 50.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("super3", UpgradeSpec { cost: 3500, additive: &[("pierce_max", 1.0)], absolute: &[], flags: &[("icebreak", true), ("laser", true)], reset_tsls: false }),
    ("super4", UpgradeSpec { cost: 4000, additive: &[("pierce_max", 1.0)], absolute: &[("attack_rate", 1.0)], flags: &[("icebreak", true), ("leadbreak", true), ("laser", true)], reset_tsls: false }),

    // Spike-o-pult
    ("spikeopult1", UpgradeSpec { cost: 250, additive: &[("attack_radius", 20.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("spikeopult2", UpgradeSpec { cost: 825, additive: &[], absolute: &[("pierce_max", 20.0)], flags: &[], reset_tsls: false }),
    ("spikeopult3", UpgradeSpec { cost: 250, additive: &[("attack_rate", -8.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("spikeopult4", UpgradeSpec { cost: 575, additive: &[], absolute: &[], flags: &[("transformed", true), ("is_spread", true)], reset_tsls: true }),

    // Beacon (drums only; storm upgrades deferred)
    ("beacon1", UpgradeSpec { cost: 500, additive: &[("attack_radius", 30.0)], absolute: &[], flags: &[], reset_tsls: false }),
    ("beacon2", UpgradeSpec { cost: 1500, additive: &[], absolute: &[], flags: &[], reset_tsls: false }),
];

/// Return the upgrade name the given path button would buy next, or None
/// if the path is fully maxed.
pub(crate) fn next_path_upgrade(
    tower_type: TowerType,
    path: u8,
    u1: bool, u2: bool, u3: bool, u4: bool,
) -> Option<&'static str> {
    let prefix = tower_type.name();
    if path == 1 {
        if !u1 {
            // e.g. "dart1"
            // We need to construct the name. Since UPGRADES is keyed by name,
            // we can look it up directly. But we need a static string.
            // The upgrade name is "{prefix}{suffix}" where suffix is "1" or "2".
            // We iterate UPGRADES to find matching entries.
            return find_upgrade_name(prefix, "1");
        }
        if !u2 {
            return find_upgrade_name(prefix, "2");
        }
        return None;
    }
    if path == 2 {
        if !u3 {
            return find_upgrade_name(prefix, "3");
        }
        if !u4 {
            return find_upgrade_name(prefix, "4");
        }
        return None;
    }
    None
}

fn find_upgrade_name(prefix: &str, suffix: &str) -> Option<&'static str> {
    // Linear scan avoids allocation. UPGRADES is small (≤32 entries).
    let expected_len = prefix.len() + suffix.len();
    UPGRADES.iter()
        .find(|(name, _)| name.len() == expected_len
              && name.starts_with(prefix)
              && name.ends_with(suffix))
        .map(|(name, _)| *name)
}

/// Check whether an upgrade can be bought (prerequisite flags).
pub(crate) fn can_upgrade(tower_type: TowerType, upgrade_name: &str, u1: bool, u2: bool, u3: bool, u4: bool) -> bool {
    let prefix = tower_type.name();
    if !upgrade_name.starts_with(prefix) {
        return false;
    }
    let suffix = &upgrade_name[prefix.len()..];
    match suffix {
        "1" => !u1,
        "2" => u1 && !u2,
        "3" => !u3,
        "4" => u3 && !u4,
        _ => false,
    }
}
